// devextreme/dist/transpiled/ui/validation/default_adapter.js
"use strict";

var Callbacks = require("../../core/utils/callbacks"),
    Class = require("../../core/class");

var DefaultAdapter = Class.inherit({
    ctor: function ctor(editor, validator) {
        var that = this;
        that.editor = editor;
        that.validator = validator;

        that.validationRequestsCallbacks = Callbacks();

        var handler = function handler() {
            that.validationRequestsCallbacks.fire();
        };

        editor.validationRequest.add(handler);

        editor.on("disposing", function () {
            editor.validationRequest.remove(handler);
        });
    },

    getValue: function getValue() {
        return this.editor.option("value");
    },

    getCurrentValidationError: function getCurrentValidationError() {
        return this.editor.option("validationError");
    },

    bypass: function bypass() {
        return this.editor.option("disabled");
    },

    applyValidationResults: function applyValidationResults(params) {
        this.editor.option({
            isValid: params.isValid,
            validationError: params.brokenRule
        });
    },

    reset: function reset() {
        this.editor.reset();
    },

    focus: function focus() {
        this.editor.focus();
    }
});

module.exports = DefaultAdapter;
