// devextreme/dist/transpiled/ui/date_box/ui.date_box.strategy.date_view.js
"use strict";

var $ = require("../../core/renderer"),
    window = require("../../core/utils/window").getWindow(),
    DateView = require("./ui.date_view"),
    DateBoxStrategy = require("./ui.date_box.strategy"),
    support = require("../../core/utils/support"),
    extend = require("../../core/utils/extend").extend,
    themes = require("../themes"),
    dateUtils = require("./ui.date_utils"),
    messageLocalization = require("../../localization/message");

var DateViewStrategy = DateBoxStrategy.inherit({

    NAME: "DateView",

    getDefaultOptions: function getDefaultOptions() {
        return extend(this.callBase(), {
            openOnFieldClick: true,
            applyButtonText: messageLocalization.format("Done")
        });
    },

    getDisplayFormat: function getDisplayFormat(displayFormat) {
        return displayFormat || dateUtils.FORMATS_MAP[this.dateBox.option("type")];
    },

    popupConfig: function popupConfig() {
        return {
            showTitle: true,
            toolbarItems: this.dateBox._popupToolbarItemsConfig(),

            defaultOptionsRules: [{
                device: function device(_device) {
                    return _device.platform === "win" && _device.version && _device.version[0] === 8;
                },
                options: {
                    showNames: true
                }
            }, {
                device: function device(_device2) {
                    return _device2.platform === "win" && _device2.phone && _device2.version && _device2.version[0] === 8;
                },
                options: {
                    animation: null
                }
            }, {
                device: function device() {
                    var currentTheme = (themes.current() || "").split(".")[0];
                    return currentTheme === "win8";
                },
                options: {
                    fullScreen: true
                }
            }, {
                device: { platform: "android" },
                options: {
                    width: 333,
                    height: 331
                }
            }, {
                device: function device(_device3) {
                    var platform = _device3.platform,
                        version = _device3.version;
                    return platform === "generic" || platform === "ios" || platform === "win" && version && version[0] === 10;
                },
                options: {
                    width: "auto",
                    height: "auto"
                }
            }, {
                device: function device(_device4) {
                    var platform = _device4.platform,
                        phone = _device4.phone;

                    return platform === "generic" && phone;
                },
                options: {
                    width: 333,
                    maxWidth: "100%",
                    maxHeight: "100%",
                    height: "auto",
                    position: {
                        collision: "flipfit flip"
                    }
                }
            }, {
                device: function device(_device5) {
                    var currentTheme = (themes.current() || "").split(".")[0];
                    return _device5.phone && currentTheme === "win10";
                },
                options: {
                    width: 333,
                    height: "auto"
                }
            }, {
                device: { platform: "ios", phone: true },
                options: {
                    width: "100%",
                    position: {
                        my: "bottom",
                        at: "bottom",
                        of: window
                    }
                }
            }]
        };
    },

    _renderWidget: function _renderWidget() {
        if (support.inputType(this.dateBox.option("mode")) && this.dateBox._isNativeType() || this.dateBox.option("readOnly")) {
            if (this._widget) {
                this._widget.$element().remove();
                this._widget = null;
            }

            return;
        }

        var popup = this._getPopup();

        if (this._widget) {
            this._widget.option(this._getWidgetOptions());
        } else {
            var element = $("<div>").appendTo(popup.$content());
            this._widget = this._createWidget(element);
        }

        this._widget.$element().appendTo(this._getWidgetContainer());
    },

    _getWidgetName: function _getWidgetName() {
        return DateView;
    },

    _getWidgetOptions: function _getWidgetOptions() {
        return {
            value: this.dateBoxValue() || new Date(),
            type: this.dateBox.option("type"),
            minDate: this.dateBox.dateOption("min") || new Date(1900, 1, 1),
            maxDate: this.dateBox.dateOption("max") || new Date(Date.now() + 50 * dateUtils.ONE_YEAR),
            onDisposing: function () {
                this._widget = null;
            }.bind(this)
        };
    }
});

module.exports = DateViewStrategy;
