// devextreme/dist/transpiled/ui/grid_core/ui.grid_core.filter_sync.js
"use strict";

var _type = require("../../core/utils/type");

var modules = require("./ui.grid_core.modules"),
    utils = require("../filter_builder/utils"),
    gridCoreUtils = require("./ui.grid_core.utils"),
    filterUtils = require("../shared/filtering"),
    customOperations = require("./ui.grid_core.filter_custom_operations");

var FILTER_ROW_OPERATIONS = ["=", "<>", "<", "<=", ">", ">=", "notcontains", "contains", "startswith", "endswith", "between"];

var FilterSyncController = modules.Controller.inherit(function () {
    var getEmptyFilterValues = function getEmptyFilterValues() {
        return {
            filterType: "include",
            filterValues: undefined
        };
    };

    var canSyncHeaderFilterWithFilterRow = function canSyncHeaderFilterWithFilterRow(column) {
        return !filterUtils.getGroupInterval(column) && !(column.headerFilter && column.headerFilter.dataSource);
    };

    var getHeaderFilterFromCondition = function getHeaderFilterFromCondition(headerFilterCondition, column) {
        if (!headerFilterCondition) {
            return getEmptyFilterValues();
        }

        var filterType,
            selectedFilterOperation = headerFilterCondition[1],
            value = headerFilterCondition[2],
            hasArrayValue = Array.isArray(value);

        if (!hasArrayValue) {
            if (!canSyncHeaderFilterWithFilterRow(column)) {
                return getEmptyFilterValues();
            }
        }

        switch (selectedFilterOperation) {
            case "anyof":
            case "=":
                filterType = "include";
                break;
            case "noneof":
            case "<>":
                filterType = "exclude";
                break;
            default:
                return getEmptyFilterValues();
        }

        return {
            filterType: filterType,
            filterValues: hasArrayValue ? value : [value]
        };
    };

    var getConditionFromFilterRow = function getConditionFromFilterRow(column) {
        var value = column.filterValue;
        if ((0, _type.isDefined)(value)) {
            var operation = column.selectedFilterOperation || column.defaultFilterOperation || utils.getDefaultOperation(column),
                filter = [column.dataField, operation, column.filterValue];
            return filter;
        } else {
            return null;
        }
    };

    var getConditionFromHeaderFilter = function getConditionFromHeaderFilter(column) {
        var selectedOperation,
            value,
            filterValues = column.filterValues;

        if (!filterValues) return null;

        if (canSyncHeaderFilterWithFilterRow(column) && column.filterValues.length === 1 && !Array.isArray(filterValues[0])) {
            column.filterType === "exclude" ? selectedOperation = "<>" : selectedOperation = "=";
            value = filterValues[0];
        } else {
            column.filterType === "exclude" ? selectedOperation = "noneof" : selectedOperation = "anyof";
            value = filterValues;
        }
        return [column.dataField, selectedOperation, value];
    };

    var updateHeaderFilterCondition = function updateHeaderFilterCondition(columnsController, column, headerFilterCondition) {
        var headerFilter = getHeaderFilterFromCondition(headerFilterCondition, column);
        columnsController.columnOption(column.dataField, headerFilter);
    };

    var updateFilterRowCondition = function updateFilterRowCondition(columnsController, column, condition) {
        var filterRowOptions,
            selectedFilterOperation = condition && condition[1];
        if (FILTER_ROW_OPERATIONS.indexOf(selectedFilterOperation) >= 0) {
            filterRowOptions = {
                filterValue: condition[2],
                selectedFilterOperation: selectedFilterOperation
            };
        } else {
            filterRowOptions = {
                filterValue: undefined,
                selectedFilterOperation: undefined
            };
        }
        columnsController.columnOption(column.dataField, filterRowOptions);
    };

    return {
        syncFilterValue: function syncFilterValue() {
            var that = this,
                columnsController = that.getController("columns"),
                columns = columnsController.getColumns();

            this._skipSyncColumnOptions = true;
            columns.forEach(function (column) {
                var filterConditions = utils.getMatchedConditions(that.option("filterValue"), column.dataField);
                if (filterConditions.length === 1) {
                    var filterCondition = filterConditions[0];
                    updateHeaderFilterCondition(columnsController, column, filterCondition);
                    updateFilterRowCondition(columnsController, column, filterCondition);
                } else {
                    column.filterValues && updateHeaderFilterCondition(columnsController, column);
                    column.filterValue && updateFilterRowCondition(columnsController, column);
                }
            });
            this._skipSyncColumnOptions = false;
        },

        init: function init() {
            if (this.getController("data").skipCalculateColumnFilters()) {
                this.syncFilterValue();
            }
        },

        _getSyncFilterRow: function _getSyncFilterRow(filterValue, column) {
            var filter = getConditionFromFilterRow(column);
            if ((0, _type.isDefined)(filter)) {
                return utils.syncFilters(filterValue, filter);
            } else {
                return utils.removeFieldConditionsFromFilter(filterValue, column.dataField);
            }
        },

        _getSyncHeaderFilter: function _getSyncHeaderFilter(filterValue, column) {
            var filter = getConditionFromHeaderFilter(column);
            if (filter) {
                return utils.syncFilters(filterValue, filter);
            } else {
                return utils.removeFieldConditionsFromFilter(filterValue, column.dataField);
            }
        },

        getFilterValueFromState: function getFilterValueFromState(state) {
            if (state.filterValue || !this.option("filterSyncEnabled")) {
                return state.filterValue || null;
            }

            var filterValue = ["and"],
                columns = state.columns;

            columns && columns.forEach(function (column) {
                var headerFilter = getConditionFromHeaderFilter(column),
                    filterRow = getConditionFromFilterRow(column);

                headerFilter && utils.addItem(headerFilter, filterValue);
                filterRow && utils.addItem(filterRow, filterValue);
            });
            return utils.getNormalizedFilter(filterValue, columns);
        },

        syncFilterRow: function syncFilterRow(column, value) {
            this.option("filterValue", this._getSyncFilterRow(this.option("filterValue"), column));
        },

        syncHeaderFilter: function syncHeaderFilter(column) {
            this.option("filterValue", this._getSyncHeaderFilter(this.option("filterValue"), column));
        },

        getCustomFilterOperations: function getCustomFilterOperations() {
            var filterBuilderCustomOperations = this.option("filterBuilder.customOperations") || [];
            return [customOperations.anyOf(this.component), customOperations.noneOf(this.component)].concat(filterBuilderCustomOperations);
        },

        publicMethods: function publicMethods() {
            return ["getCustomFilterOperations"];
        }
    };
}());

var DataControllerFilterSyncExtender = {
    skipCalculateColumnFilters: function skipCalculateColumnFilters() {
        var filterSyncEnabledValue = this.option("filterSyncEnabled");
        return filterSyncEnabledValue === "auto" ? this.option("filterPanel.visible") : filterSyncEnabledValue;
    },

    _calculateAdditionalFilter: function _calculateAdditionalFilter() {
        var that = this;

        if (that.option("filterPanel.filterEnabled") === false) {
            return that.callBase();
        }

        var filters = [that.callBase()],
            columns = that.getController("columns").getColumns(),
            filterValue = that.option("filterValue");

        if (that.skipCalculateColumnFilters()) {
            var currentColumn = that.getController("headerFilter").getCurrentColumn();
            if (currentColumn && filterValue) {
                filterValue = utils.removeFieldConditionsFromFilter(filterValue, currentColumn.dataField);
            }
        }
        var customOperations = that.getController("filterSync").getCustomFilterOperations(),
            calculatedFilterValue = utils.getFilterExpression(filterValue, columns, customOperations, "filterBuilder");
        if (calculatedFilterValue) {
            filters.push(calculatedFilterValue);
        }

        return gridCoreUtils.combineFilters(filters);
    },

    _parseColumnInfo: function _parseColumnInfo(fullName) {
        var matched = fullName.match(/columns\[([0-9]*)\]\.(.*)/);
        return matched ? { index: matched[1], changedField: matched[2] } : null;
    },

    optionChanged: function optionChanged(args) {
        switch (args.name) {
            case "filterValue":
                this._applyFilter();
                this.skipCalculateColumnFilters() && this.getController("filterSync").syncFilterValue();
                args.handled = true;
                break;
            case "filterSyncEnabled":
                args.handled = true;
                break;
            case "columns":
                if (this.skipCalculateColumnFilters()) {
                    var columnInfo = this._parseColumnInfo(args.fullName),
                        column = void 0,
                        filterSyncController = this.getController("filterSync");
                    if (columnInfo && !filterSyncController._skipSyncColumnOptions) {
                        filterSyncController._skipSyncColumnOptions = true;
                        if (["filterValues", "filterType"].indexOf(columnInfo.changedField) !== -1) {
                            column = this.getController("columns").getColumns()[columnInfo.index];
                            filterSyncController.syncHeaderFilter(column);
                        } else if (["filterValue", "selectedFilterOperation"].indexOf(columnInfo.changedField) !== -1) {
                            column = this.getController("columns").getColumns()[columnInfo.index];
                            filterSyncController.syncFilterRow(column, column.filterValue);
                        }
                        filterSyncController._skipSyncColumnOptions = false;
                    }
                }
                break;
            default:
                this.callBase(args);
        }
    }
};

var ColumnHeadersViewFilterSyncExtender = {
    _isHeaderFilterEmpty: function _isHeaderFilterEmpty(column) {
        if (this.getController("data").skipCalculateColumnFilters()) {
            return !utils.filterHasField(this.option("filterValue"), column.dataField);
        }

        return this.callBase(column);
    },

    _needUpdateFilterIndicators: function _needUpdateFilterIndicators() {
        return !this.getController("data").skipCalculateColumnFilters();
    },

    optionChanged: function optionChanged(args) {
        if (args.name === "filterValue") {
            this._updateHeaderFilterIndicators();
        } else {
            this.callBase(args);
        }
    }
};

module.exports = {
    defaultOptions: function defaultOptions() {
        return {
            /**
             * @name GridBaseOptions.filterValue
             * @publicName filterValue
             * @type Filter expression
             * @default null
             */
            filterValue: null,

            /**
             * @name GridBaseOptions.filterSyncEnabled
             * @publicName filterSyncEnabled
             * @type boolean
             */
            filterSyncEnabled: "auto"
        };
    },
    controllers: {
        filterSync: FilterSyncController
    },
    extenders: {
        controllers: {
            data: DataControllerFilterSyncExtender
        },
        views: {
            columnHeadersView: ColumnHeadersViewFilterSyncExtender
        }
    }
};
